### PR TITLE
OOP - Milestone 3: game_logic

### PR DESCRIPTION
**Pull request by Gnwin**
## Milestone 3: Game_logic  

https://repl.it/@CodeVantage/OOP#main.rb

Use the above link as a temporary preview for the project.

In this branch, The game logic was programmed following instructions using:
- classes
- objects
- accessor methods
- encapsulation
- modules
- procs

 The game has features such as:
- Singleplayer(player2 is the computer) 
- Multiplayer(between two humans)

### CODE REVIEW

**Comment by Gnwin** [comment]
https://repl.it/@CodeVantage/OOP#main.rb

Use the above link as a temporary preview for the project.
Please advise on any changes.

**Review by enaburekhan — COMMENTED** [review]
# Status: Invalid
Friendly reminder: You should make sure that your repo looks professional before review.
All linter checks must be passed before review. This Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready!

You can drop your questions here and ping me on slack @ Eric Enaburekhan to respond.
Happy Coding!
Eric Enaburekhan

File: bin/main [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+      @slot = gets.chomp.to_i
+      break if board[@slot] == '-'
+    end
+    @slot
+  end
+end
+
+class Game
+  include Display
+  attr_accessor :val
```
Comment (bini-i) [inline bin/main]: Try to use attr_reader otherwise, this exposes your attributes to unknow changes from outside your class.

**Review by bini-i — CHANGES_REQUESTED** [review]
# Status: Changes Required :yellow_circle: 
### Hello  @vicmaburrito and @Gnwin :wave: 

### Your project is very nice :clap: :muscle: . The game works and I like that you have both single and multiplayer options, but you still have somethings to work on and improve.

- [ ] All game-logic related resources (classes and/or modules) are kept in the lib directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)). currently you have everything under main. Kindly apply the requirement.
- [ ] No `puts/print` and `gets.chomp` outside `main.rb` file or in game logic related resources(classes)

#### Kindly collaborate on the project with equal contribution to the commit history.

#### If you have any question kindly reach me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (TSE)
> GitHub: [@bini-i](https://github.com/bini-i)
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

File: lib/tic_tac_toe.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    puts '4 | 5 | 6'
+    puts '---------'
+    puts '7 | 8 | 9'
+    puts "
+		"
+  end
+
+  def choose(name0, arr0)
+    puts "its #{name0}'s turn to choose a number from #{arr0}"
+  end
```
Comment (BrianSammit) [inline lib/tic_tac_toe.rb]: These two methods `display_guide` and `choose` have `puts`. Please move them inside the `main.rb` file :hammer_and_wrench:

File: lib/tic_tac_toe.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    end
+  end
+
+  def display_slots
+    puts "#{Board.squares[1]}  #{Board.squares[2]}  #{Board.squares[3]}"
+    puts "#{Board.squares[4]}  #{Board.squares[5]}  #{Board.squares[6]}"
+    puts "#{Board.squares[7]}  #{Board.squares[8]}  #{Board.squares[9]}"
+    puts "
+    "
+  end
```
Comment (BrianSammit) [inline lib/tic_tac_toe.rb]: This method `display_slots` have `puts`. Kindly move it to the `main.rb` file. :hammer_and_wrench:

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :recycle: :red_circle: 
-----------
Hello team!! :wave: you are doing an amazing job on this milestone.
However, there are a few changes to make :hammer_and_wrench: see them below :eyes: :point_down: 
---------

- [ ] You have `puts` outside the main file, please move this code inside the main file. Remember that "No puts/print and gets.chomp outside main.rb file". See my inline comments to know what lines of code I'm talking about. 
no_comment.

- [ ] Improve your commit messages. 
- Remember that you should write your commit messages in the imperative. [Here](https://github.com/trein/dev-best-practices/wiki/Git-Commit-Best-Practices#formatting-rules) you can read about the formating rule for the commit messages. 
- You have often repeated the same commit messages. [Here](https://docs.github.com/en/github/committing-changes-to-your-project/changing-a-commit-message) you can read how to change a commit message. To not commit the same message, make the changes, and then follow the instructions from the link, you gonna create a new commit that overrides the old commit. 

You don't need to make any changes this time with the commit messages but keep this instruction in mind for the next commits. 

-------------
Please make the required changes above :point_up_2: and send it for another review. 
HAPPY CODING!!

MB5W4D

**Review by smcommits — APPROVED** [review]
# Status: Approved :+1: 
Hello, team!

**Great job of implementing the changes requested by previous TSEs. Your application looks great. 
This PR is now approved. You can go ahead and merge the branch. Congratulations and good luck with your next project.** 
# 
Any further questions? Please reach out to me on [slack](https://microverse-students.slack.com/team/U017N7LLZ40)

**Pull request merged**